assets/arc-0052/contextual.api.crypto.ts

```diff
@@ -0,0 +1,292 @@
+            // Could be an implementation choice. 
+            // Example: 
+            // const nodeScalar: Uint8Array = derived.subarray(0, 32)
+            // const nodePublic: Uint8Array = crypto_scalarmult_ed25519_base_noclamp(nodeScalar)
+            // const nodeCC: Uint8Array = derived.subarray(64, 96)
+
+            // // [Public][ChainCode]
+            // const extPub: Uint8Array = new Uint8Array(Buffer.concat([nodePublic, nodeCC]))
+            // const publicKey: Uint8Array = deriveChildNodePublic(extPub, bip44Path[4]).subarray(0, 32)
+
+            derived = deriveChildNodePrivate(derived, bip44Path[4])
+
+        const scalar = derived.subarray(0, 32) // scalar == pvtKey
+        return isPrivate ? scalar : crypto_scalarmult_ed25519_base_noclamp(scalar)
+    }
+
+    /**
+     * 
+     * 
+     * @param context - context of the key (i.e Address, Identity)
+     * @param account - account number. This value will be hardened as part of BIP44
+     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
+     * @returns - public key 32 bytes
+     */
+    async keyGen(context: KeyContext, account:number, keyIndex: number): Promise<Uint8Array> {
+        await ready // libsodium
+
+        const rootKey: Uint8Array = fromSeed(this.seed)
+        const bip44Path: number[] = GetBIP44PathFromContext(context, account, keyIndex)
+
+        return await this.deriveKey(rootKey, bip44Path, false)
+    }
+
+    /**
+     * Ref: https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.6
+     * 
+     *  Edwards-Curve Digital Signature Algorithm (EdDSA)
+     * 
+     * @param context - context of the key (i.e Address, Identity)
+     * @param account - account number. This value will be hardened as part of BIP44
+     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
+     * @param data - data to be signed in raw bytes
+     * @param metadata - metadata object that describes how `data` was encoded and what schema to use to validate against
+     * 
+     * @returns - signature holding R and S, totally 64 bytes
+     * */ 
+    async signData(context: KeyContext, account: number, keyIndex: number, data: Uint8Array, metadata: SignMetadata): Promise<Uint8Array> {
+        // validate data
+        const result: boolean | Error = this.validateData(data, metadata)
+        
+        if (result instanceof Error) { // decoding errors
+            throw result
+        }
+
+        if (!result) { // failed schema validation
+            throw ERROR_BAD_DATA
+        }
+        
+        await ready // libsodium
+
+        const rootKey: Uint8Array = fromSeed(this.seed)
+        const bip44Path: number[] = GetBIP44PathFromContext(context, account, keyIndex)
+        const raw: Uint8Array = await this.deriveKey(rootKey, bip44Path, true)
+
+        const scalar = raw.slice(0, 32);
+        const c = raw.slice(32, 64);
```

Comment:
I believe there is a bug here. c is supposed to be the raw.slice(32, 64), but deriveKey will always return raw as a 32 byte length Uin8Array. You can see this inside deriveKey function on line 97 that we take scalar = derived.subarray(0, 32).


When raw.slice(32, 64) gets run here on a size 32 array, rather than fail it will simply return a 0 length Uint8Array. Later on when concatenating c and data to produce the first hash for r (currently line 156), c is empty and as a result only data gets hashed.

I'm a little unsure on how to proceed. According to the [RFC 8032 section 3.3](https://datatracker.ietf.org/doc/html/rfc8032#page-8), also implemented [noble-curves](https://github.com/paulmillr/noble-curves/blob/6a85252dc3f9d8bef7e407177c6c81a3a67b185b/src/abstract/edwards.ts#L449) (c is called "prefix" and is calculated on line 426) you would grab your private key, hash it to get 64 bytes. The first 32 bytes would make up the scalar to calculate the scalar*B = publicKey (which we do). And then the last 32 bytes would be used for this.

Should we change deriveKey so instead of isPrivate (true/false), it could have three modes: get scalar, get publicykey and get raw (full 64 bytes)?

Reply:
Btw, for those confused, the code still works, it's just that R (the first 32 bytes of the signature) will not be produced correctly. The problem is that r will be calculable as a function of the message data, a known value. If you know r you can take use the other half the signature, S, and calculate the private key as such: (S - r)*h^-1 = k

Reply:
updated